### DISCUSSION TITLE
Larger than memory joins

### DISCUSSION THREAD

**Discussion by CerebralMastication** [duckdb/duckdb#3820] [Q&A]
Hey all, I've been building some examples to teach myself how to best use DuckDB in my Python workflows. Such a great tool and very well engineered. 

I'm running into memory issues with large joins, however. And since I can't find much discussion online about the issue I wonder if I'm just doing something naive and making life hard on myself. 

What I'm trying to do is take ~3 billion records over in S3 stored in parquet files and join them to another table with 100k records. This is a many to many join so my result would have ~25 billion records. There are 8 fields in my large table and the join adds a 9th field to the results. I am not returning the results to Python, but rather have tried both a `CREATE TABLE AS` query and a `COPY ( ) TO 's3::/bucket/' (FORMAT 'parquet')` 

If I just naively do the join my JupyterLab kernel crashes. A dead give away that I've run out of RAM. Doing some research I learned that DuckDB may not be aware of the amount of memory to use, so I give it a hint and set 

`PRAGMA memory_limit='16GB';` 

That does change the behavior because JupyterLab kernel no longer crashes, but I do get a different memory error: `RuntimeError: Out of Memory Error: could not allocate block of 262144 bytes`

Is there an approach here that I am missing? Or should I be doing this differently?

**Comment by Mytherin** [comment]
Are you running DuckDB using an in-memory database, or are you using a disk-backed database? By default DuckDB will not try to offload data to disk if it is running in in-memory mode, where-as when using a disk-backed database DuckDB will try to offload data to the `mydb.db.tmp` directory. You can manually specify a directory to off-load to as well using `SET temp_directory='/tmp/mytmp.duck';`.

**Reply by CerebralMastication** [reply]
Oh.. yeah.. that makes sense. I was totally missing that DuckDB was in memory by default unless I specify a backing file. Though that should be kind of obvious to me, I was overlooking that. Thank you for the nudge. That alone may solve a number of mysteries I was having.

**Reply by CerebralMastication** [reply]
Mark, 

I'm now setting up my connection like this:
```
con = duckdb.connect(database='test.duck')
con.execute("""
    SET temp_directory='tmp.duck';
    PRAGMA memory_limit='32GB';
    PRAGMA threads=16; """
) 
```
my large file I want to join to is currently stored as a number of parquet files in S3, and I'm on DuckDB 0.3.4 so no S3 globbing. To work around that I do the ol' "arrow dataset" work around:

```
big_dataset = ds.dataset("s3://roc_bucket/large_boi/")
con.register("big", big_dataset)
```
I do the same logic for the smaller dataset, which I register as `small`

Then I run the following query:

```

sql = """
create table test_big_join as  

    SELECT A, B, C, D, 
        sum(amount) as amount
    FROM small s
        JOIN  big b on b.id = s.id
    GROUP BY 
        A, B, C, D

;
"""
%time con.execute(sql)
```

when I run that it runs for less than a minute then crashes my JuypterLab kernel. I was watching the size of the `*.duck` files I specified and they never changed size:

```
bash-4.2$ ll *.duck
-rw-r--r-- 1 jal jal 12288 Jun 10 16:40 mytmp.duck
-rw-r--r-- 1 jal jal 12288 Jun 10 17:01 test.duck
```

So I'm fairly perplexed. But it does have me wondering if my use of that Arrow dataset work around for globbing is somehow making this not work well. 

as a test I'll get a list of the parquet files and try listing them explicitly in the duck query, instead of using the arrow dataset. If you see something obvious that I'm missing please point it out.

**Reply by Mytherin** [reply]
How many distinct values are there in `A, B, C, D`? We don't support offloading of large aggregate tables to disk (yet).

**Reply by CerebralMastication** [reply]
ahhhh.. this is probably where I'm running into trouble. A has 40,000 unique value, B has 5000, C has 5, D has 1000... so yeah, lots of combinations. 

sounds like I need an explicit chunking strategy. That's how I have been handling this data in the past I just thought I'd try just throwing at at DuckDB and seeing how it plays. 

thank you very much for your fast response. Really appreciated.

**Comment by orzom411** [comment]
I would suggest working to segment your data into workable sets, DuckDB is particularly flexible but like all management systems there are limitations.  I've found a divide and conquer a reliable way of managing memory, actually it's become my first line of offense as it allows me to thread out concurrent operations.  Generally speaking it pay dividends in terms of overall processing as I can consume an appliance's full bandwidth in terms of processing power, as well as keeps my memory footprint in check.

In your case your 100k data, if you took a simple slice of say 1k do you subsume your allotted memory?  Remember, letting the underlying engine decide what to swap out may not be as advantageous as knowledge you posses about the data being processed.  Just a thought, hope it helps you look at the memory footprint as a way to parallel your workflow.

**Reply by CerebralMastication** [reply]
excellent philosophical nudge. When you do "segmenting" (I often think of that as chunking) do you sort of hand roll the chunk size or are there tools that help manage that process? I don't mind highly controlling that process of chunking myself but it's totally an optimization and Uncle Donald once warned me about the problem with premature optimization :)

**Reply by orzom411** [reply]
I'm a baker, hands on first then if I can see something useful to pattern after, well I'll mint that instead.  Funny thing is when you start chunking you often find there was a "natural" chunk built into the data that drove your summations.  Maybe look at your group by statement, might it essentially point out where your chunks may be most readily available?  I also like to review index usage, systems often lay before it's administrators exactly where to partition there data, in the form of what questions are being asked.  You maybe able to make the connection into chunks just by tossing a sample set into excel and rolling it around for half hour or so.  That is bake it just a bit till you tease out what question to ask that gives you a reasonable chunk size that won't blowout your memory nor take oodles of time to process.

One case used unique identifiers and I simply split them into segments, based on there binary representation.  The DB is happy, seeks for each chunk, my memory footprint remained manageable and as a wonderful side I get to report back to administrators progress.

Hope that helps.

**Comment by rsund** [comment]
Similar to #3554

**Comment by vtslab** [comment]
I ran into the same issue where a large joined table was not spilled to disk and the OOM occurred. I my case it was a join of 4 tables. I managed to do the join in the end by first creating a joined table with only the primary keys of all tables and then a COPY TO clause of this table joined with the individual tables. Apparently, this induces a streaming behavior that does not need the entire joined table in memory. I could even imagine that a query planner would take this approach by default!

**Reply by lnkuiper** [reply]
Once #4189 is merged DuckDB will support larger-than-memory joins.

I will continue to optimize this implementation in the coming weeks/months.
